dynamic-floating-group.lisp

```diff
@@ -0,0 +1,461 @@
+  (alexandria:appendf (dyn-float-group-dyn-order group)
+                      (list (make-instance 'window+ :window window :drift nil)))
+  (re-tile group))
+
+(defmethod stumpwm:group-delete-window
+    ((group dyn-float-group)
+     (window stumpwm::float-window))
+  (declare (ignore window))
+  (stumpwm::%float-focus-next group)
+  (sync-dyn-order group)
+  (re-tile group))
+
+(defmethod stumpwm:group-button-press
+    ((group dyn-float-group)
+     button x y (window stumpwm::float-window))
+  ;; Free the window if it's pressed at the boarder or with
+  ;; *float-window-modifier*.
+  (let ((xwin (stumpwm:window-xwin window)))
+    (multiple-value-bind (relx rely same-screen-p child state-mask)
+        (xlib:query-pointer (stumpwm::window-parent window))
+      (declare (ignore relx rely same-screen-p child))
+      (when (or
+             (< x (xlib:drawable-x xwin))
+             (> x (+ (xlib:drawable-width xwin)
+                     (xlib:drawable-x xwin)))
+             (< y (xlib:drawable-y xwin))
+             (> y (+ (xlib:drawable-height xwin)
+                     (xlib:drawable-y xwin)))
+             (intersection (stumpwm::float-window-modifier)
+                           (xlib:make-state-keys state-mask)))
+        (free-window window group))))
+  (call-next-method))
+
+;;;
+
+(defun sync-dyn-order (&optional (group (stumpwm:current-group)))
+  "Recall that a window+ is a window with some extra data.
+Dynamic-floating-group works by operating on the DYN-ORDER, i.e.
+the list of window+ of the group.
+
+This function ensures that both the ordinary window list and the
+window+ list are in sync. It then ensures that the drifting windows
+stay in the beginning of the DYN-ORDER. After doing so, it
+ensures the ordinary window list respects the order of
+DYN-ORDER."
+  (assert (dyn-float-group-p group) ()
+          "Expected GROUP ~A to be of type DYN-FLOAT-GROUP." group)
+  (symbol-macrolet ((dyn-order (dyn-float-group-dyn-order group)))
+
+    ;; Sync (w+) according to (w).
+    ;;
+    ;; If window W does not have a corresponding W+ in the
+    ;; dyn-order, make one for it.
+    (loop for w in (stumpwm::group-windows group)
+          unless (member w (mapcar #'window+-window dyn-order))
+          do (push (make-instance 'window+ :window w :drift nil) dyn-order))
+    ;; If window W+ does not correspond to a window of GROUP,
+    ;; delete W+ from the dyn-order.
+    (loop for w+ in dyn-order
+          unless (member (window+-window w+) (stumpwm::group-windows group))
+          do (alexandria:deletef dyn-order w+))
+
+    ;; Make the drifting windows on top of the stack.
+    (setf dyn-order
+          (append (remove-if (lambda (dyno)
+                               (equal nil (window+-drift dyno)))
+                             dyn-order)
+                  (remove-if (lambda (dyno)
+                               (equal t (window+-drift dyno)))
+                             dyn-order)))
```

Comment:
We can use sort for this. See the following snippet

```lisp
* (sort (list t nil t nil t) (lambda (x y) x))
(T T T NIL NIL)
```

So something like
```lisp
;; We define a predicate as a top level form outside this function
(defun drifting-first (win-a win-b)
  (declare (ignore win-b))
  (window+-drift win-a))
...  
    ;; Make the drifting windows on top of the stack.
    (setf dyn-order
      (sort (copy-list dyn-order) #'drifting-first))
```

should work. A common CL gotcha is that sort destructively modifies the list which is why I'm adding the copy-list.

P.D.

The current approach could be made shorter by removing the surrounding (equal ..) form and using remove-if-not for the first filter. something like:

```lisp
    ;; Make the drifting windows on top of the stack.
    (setf dyn-order
          (append (remove-if-not #'window+-drift dyn-order)
                  (remove-if #'window+-drift dyn-order)))
```

But I think sort captures the intent of the code better.

Reply:
Nice fu! Done.